data-prepper-plugins/cloudwatch-metrics-source/src/main/java/org/opensearch/dataprepper/plugins/source/CloudwatchMetricsWorker.java

```diff
@@ -0,0 +1,198 @@
+        this.bufferAccumulator = bufferAccumulator;
+        this.cloudWatchClient = cloudWatchClient;
+        this.cloudwatchMetricsSourceConfig = cloudwatchMetricsSourceConfig;
+        this.dimensionCollection = dimensionCollection;
+        this.pluginMetrics = pluginMetrics;
+        this.sourceCoordinator = sourceCoordinator;
+        final List<String> namespaceMetricNames = cloudwatchMetricsSourceConfig.getNamespacesListConfig().stream()
+                .map(NamespacesListConfig::getNamespaceConfig).flatMap(namespaceConfig -> namespaceConfig.getMetricDataQueriesConfig()
+                        .stream().map(metricName -> namespaceConfig.getName() + "@" + metricName.getMetricsConfig().getName())).collect(Collectors.toList());
+        this.partitionCreationSupplier = new CloudwatchPartitionCreationSupplier(namespaceMetricNames);
+    }
+
+    void startProcessingCloudwatchMetrics(final int waitTimeMillis){
+        final Optional<SourcePartition<CloudwatchSourceProgressState>> objectToProcess = sourceCoordinator.getNextPartition(partitionCreationSupplier);
+
+        if (objectToProcess.isEmpty()) {
+            try {
+                Thread.sleep(waitTimeMillis);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+            return;
+        }
+        final String nameSpaceName = objectToProcess.get().getPartitionKey().split("@")[0];
+        final String metricName = objectToProcess.get().getPartitionKey().split("@")[1];
+        try {
+            processCloudwatchMetrics(nameSpaceName,metricName);
+            sourceCoordinator.completePartition(objectToProcess.get().getPartitionKey());
+        } catch (final PartitionNotOwnedException | PartitionNotFoundException | PartitionUpdateException e) {
+            LOG.warn("Cloud watch source received an exception from the source coordinator. There is a potential for duplicate data from {}, giving up partition and getting next partition: {}",objectToProcess.get().getPartitionKey() , e.getMessage());
+            sourceCoordinator.giveUpPartitions();
+        }
+    }
+
+    protected void processCloudwatchMetrics(final String nameSpaceName, final String metricName) {
+        for(NamespacesListConfig namespacesListConfig : cloudwatchMetricsSourceConfig.getNamespacesListConfig().stream()
+                .filter(obj -> obj.getNamespaceConfig().getName().equals(nameSpaceName)).collect(Collectors.toList())) {
+            dimensionCollection.clear();
+            for (MetricDataQueriesConfig metricDataQueriesConfig : namespacesListConfig.getNamespaceConfig().getMetricDataQueriesConfig().stream().
+                    filter(obj -> obj.getMetricsConfig().getName().equals(metricName)).collect(Collectors.toList())) {
+
+                for (DimensionsListConfig dimensionsListConfig : metricDataQueriesConfig.getMetricsConfig().getDimensionsListConfigs()) {
+
+                    dimensionCollection.add(Dimension.builder()
```

Comment:
You could build this object on start-up and then keep it in a map for future re-use. But, you'd need to not modify it.

```
private Map<String, Collection<Dimension>> metricNameToDimensions;
```

Reply:
this has been fixed.